Sources/Fuzzilli/Core/ProgramBuilder.swift

```diff
@@ -702,18 +702,45 @@ public class ProgramBuilder {
         //   Otherwise copy the whole block including its content
         var requiredInputs = VariableSet()
 
+        // A Set of variables that have yet to be included in the slice
+        var remainingInputs = VariableSet()
+
+        // A stack of contexts that are required by the instruction in the slice
+        var requiredContextStack = [Context.empty]
+
         // Helper function to add an instruction, or possibly multiple instruction in the case of blocks, to the slice.
         func add(_ instr: Instruction, includeBlockContent: Bool = false) {
             guard !slice.contains(instr.index) else { return }
 
             func internalAdd(_ instr: Instruction) {
+                if !remainingInputs.isEmpty && instr.hasOutputs {
+                    for output in instr.outputs {
+                        if remainingInputs.contains(output) {
+                            remainingInputs.remove(output)
+                        }
+                    }
+                }
+
+                if !remainingInputs.isEmpty && !instr.innerOutputs.isEmpty {
+                    for innerOutput in instr.innerOutputs {
+                        if remainingInputs.contains(innerOutput){
+                            remainingInputs.remove(innerOutput)
+                        }
+                    }
+                }
+                
                 requiredInputs.formUnion(instr.inputs)
+                remainingInputs.formUnion(instr.inputs)
+                var currentContext = requiredContextStack.popLast()!
+                currentContext.formUnion(instr.op.requiredContext)
+                requiredContextStack.append(currentContext)
                 slice.insert(instr.index)
             }
 
             if instr.isBlock {
                 let group = BlockGroup(around: instr, in: source)
-                for instr in includeBlockContent ? group.includingContent() : group.excludingContent() {
+                let instructions = includeBlockContent ? group.includingContent() : group.excludingContent()
+                for instr in instructions.reversed() {
```

Comment:
I'm also wondering if something breaks since this logic bypasses the context stack handling (starting in lines 791), and could potentially add multiple blocks (e.g. entire switch-case, class definitions, etc.) without modifying the stack. But it's a bit hard to wrap my head around what would actually happen...
Maybe it'd be easier to move this logic into the loop (starting in line 757) to reason about what happens. But I'll think about this again, maybe there's an easier/more obviously correct way.

Reply:
> Maybe it'd be easier to move this logic into the loop (starting in line 757) to reason about what happens. But I'll think about this again, maybe there's an easier/more obviously correct way.

So one of the challenges with doing that is then we would have to duplicate the handling for the first instruction that we add to the slice (the one that's outside the loop).

Also, if there are scenarios that you want me to try out then let me know and I'll whip up some testcases for it.

Reply:
So would the following testcase work currently:
```
v1 <- BeginFunctionDefinition
  BeginForLoop
    Break
  EndForLoop
EndFunctionDefinition
CallFunction v1
```
And start at `CallFunction`. IIUC correctly, this would happen:
- It takes CallFunction
- It pushes a new entry onto the scope stack for `EndFunctionDefinition` and again for `EndForLoop`
- It removes one entry for `BeginForLoop`
- It takes the entire block content at `BeginFunctionDefinition`, and so adds the `.loop` requirement into the current entry of the stack for `Break`, but doesn't remove that requirement when adding the `BeginForLoop`
- It removes one entry for `BeginFunctionDefinition` and now adds the `.loop` requirement to the remaining context (see https://github.com/googleprojectzero/fuzzilli/pull/256/files#r709268658)
- So at the end, the requiredContext is still `[.loop]`, which is incorrect and prevents the splicing from succeeding

Reply:
This splice in this scenario would be:
```
BeginPlainFunctionDefinition, inouts_: [v0, v1, v2, v0]
LoadInteger, inouts_: [v3, v1]
LoadInteger, inouts_: [v4, v2]
LoadInteger, inouts_: [v5, v3]
BeginFor, inouts_: [v3, v4, v5, v6, v4]
Break, inouts_: [v5]
EndFor, inouts_: [v6]
EndPlainFunctionDefinition, inouts_: [v7]
CallFunction, inouts_: [v0, v7, v8]
```

So it does remove the `.loop` context and at the end of the splice we are left with `.script` as the required context.

Reply:
After making the latest changes (i.e. merging of states) this is the one testcase that currently fails. I don't believe it has anything to do with the merging logic and it's more to do with how we handle adding entire blocks in `add`.

Reply:
Ok, yeah I think this testcase should fail once the merging logic works correctly, but with `add` not taking blocks into account.
So I agree that `add` is the problem, and we probably need to somehow force the block handling to happen in the outer loop. :/ I'll think about how to best achieve that

Reply:
I've duplicated the block handling in a helper function that is now being called from `add`. It's a naive solution that works but there's room for improvement here.